File: packages/node-sdk/src/facade.ts
/**
 * React-hooks-style façade over the Neon3 SDK.
 *
 * This is the public "easy mode" API. It wraps `NeonApp` and `ObservableStore`
 * so a counter demo reads:
 *
 * ```ts
 * import { start, mount, state, on } from "@neon3/sdk";
 *
 * const app = await start({ mode: "windowed" });
 * await mount("counter.nui");
 * const count = state(0);
 *
 * on("btn.increment")(() => {
 *   count.value++;
 * });
 *
 * await app.run();
 * ```
 *
 * The lower-level `NeonApp` is still available as `app.neon` inside the
 * context if you need fine-grained control.
 */

import { NeonApp } from "./app.js";
import { ObservableStore } from "./store.js";

// ---------------------------------------------------------------------------
// Global "current app" context.
//
// The module-level `mount`/`state`/`on` helpers operate on the most recently
// started app. Call `start()` to set it; call `app.dispose()` to clear it.
// ---------------------------------------------------------------------------

let currentApp: FacadeApp | null = null;

export function getCurrentApp(): FacadeApp {
  if (!currentApp) {
    throw new Error(
      "No active Neon3 app. Call `await start(...)` before mount()/state()/on().",
    );
  }
  return currentApp;
}

// ---------------------------------------------------------------------------
// State<T>: a readable/writable scalar that auto-publishes on assignment.
// ---------------------------------------------------------------------------

export class State<T = unknown> {
  readonly key: string;

  constructor(key: string, private readonly app: FacadeApp) {
    this.key = key;
  }

  get value(): T {
    const raw = this.app.store.value(this.key).get();
    // Unwrap the wire envelope { kind, value } into the raw value.
    if (raw && typeof raw === "object" && "kind" in raw && "value" in raw) {
      return (raw as unknown as { value: T }).value;
    }
    return raw as T;
  }

  set value(v: T) {
    this.app.store.value(this.key).set(v);
    void this.app.publishNow();
  }
}

// ---------------------------------------------------------------------------
// Sub-namespaces.
// ---------------------------------------------------------------------------

class ShaderNamespace {
  private readonly handlers = new Map<number, (payload: number[]) => void>();

  constructor(private readonly app: FacadeApp) {}

  /** Register a handler for a specific shader event_id. */
  on(eventId: number, handler: (payload: number[]) => void): void {
    this.handlers.set(eventId, handler);
  }

  /** Internal: dispatch one decoded shader event. */
  dispatch(eventId: number, payload: number[]): void {
    const h = this.handlers.get(eventId);
    if (h) {
      try {
        h(payload);
      } catch {
        // A user handler must not kill the dispatch loop.
      }
    }
  }
}

class RendererNamespace {
  private lastFlush = 0;
  private pending: number[][] | null = null;

  constructor(private readonly app: FacadeApp) {}

  /** Upload 1..10 rows of vec4 to `view.extras[0..9]`. Throttled to 60 Hz. */
  viewExtras(...rows: number[][]): void {
    if (rows.length < 1 || rows.length > 10) {
      throw new Error(`view_extras expects 1..10 rows, got ${rows.length}`);
    }
    for (let i = 0; i < rows.length; i++) {
      if (rows[i].length !== 4) {
        throw new Error(`row ${i} must have 4 floats, got ${rows[i].length}`);
      }
    }
    this.pending = rows.map((r) => [...r]);
    const now = performance.now();
    if (now - this.lastFlush >= 1000 / 60) {
      void this.flush();
    }
  }

  async flush(): Promise<void> {
    if (!this.pending) return;
    const rows = this.pending;
    this.pending = null;
    this.lastFlush = performance.now();
    if (this.app.neon.render) {
      await this.app.neon.render.setViewExtras(rows);
    }
  }
}

class AnimNamespace {
  constructor(private readonly app: FacadeApp) {}

  private renderClient() {
    if (!this.app.neon.render) {
      throw new Error("Capability not available: wgpu.ui.animation.control.v1");
    }
    return this.app.neon.render;
  }

  pause(nodePath: string): Promise<unknown> {
    return this.renderClient().animationPause(nodePath);
  }
  resume(nodePath: string): Promise<unknown> {
    return this.renderClient().animationResume(nodePath);
  }
  cancel(nodePath: string): Promise<unknown> {
    return this.renderClient().animationCancel(nodePath);
  }
  seek(nodePath: string, progress: number): Promise<unknown> {
    return this.renderClient().animationSeek(nodePath, progress);
  }
}

/** A rendered PNG returned by `app.surface.render(...)`. */
export class Png {
  constructor(readonly path: string) {}

  save(dest: string): Promise<void> {
    return import("node:fs/promises").then((fs) => fs.copyFile(this.path, dest));
  }

  bytes(): Promise<Buffer> {
    return import("node:fs/promises").then((fs) => fs.readFile(this.path));
  }
}

class SurfaceNamespace {
  constructor(private readonly app: FacadeApp) {}

  /** Render a NUI flow offscreen and return a PNG handle. */
  async render(
    nui: string,
    options: { size?: [number, number]; surfaceId?: string } = {},
  ): Promise<Png> {
    if (!this.app.neon.render) throw new Error("render client not available (offline?)");
    const [w, h] = options.size ?? [1280, 720];
    const sid = `${options.surfaceId ?? "facade-offscreen"}-${crypto.randomUUID().slice(0, 8)}`;
    const os = await import("node:os");
    const path = await import("node:path");
    const outPath = path.join(os.tmpdir(), `neon3-${crypto.randomUUID().slice(0, 8)}.png`);
    const surface = await this.app.neon.render.openSurface({
      session_id: this.app.neon.origin,
      surface_id: sid,
      kind: "screen_ui",
      size: { width: w, height: h },
      buffer_count: 1,
    } as never);
    await this.app.neon.ui.mountFlow(nui);
    await surface.savePng(outPath);
    return new Png(outPath);
  }
}

// ---------------------------------------------------------------------------
// FacadeApp: the high-level app object returned by start().
// ---------------------------------------------------------------------------

export class FacadeApp {
  readonly store: ObservableStore;
  readonly shader: ShaderNamespace;
  readonly renderer: RendererNamespace;
  readonly anim: AnimNamespace;
  readonly surface: SurfaceNamespace;

  private stateCounter = 0;

  private constructor(readonly neon: NeonApp) {
    this.store = neon.store ?? new ObservableStore();
    if (!neon.store) neon.store = this.store;
    this.shader = new ShaderNamespace(this);
    this.renderer = new RendererNamespace(this);
    this.anim = new AnimNamespace(this);
    this.surface = new SurfaceNamespace(this);
  }

  static async boot(options: Parameters<typeof NeonApp.start>[0] = {}): Promise<FacadeApp> {
    const neon = await NeonApp.start(options);
    const app = new FacadeApp(neon);
    currentApp = app;
    return app;
  }

  async mount(source: string): Promise<unknown> {
    return this.neon.ui.mountFlow(source);
  }

  state<T>(initial: T, node?: string): State<T> {
    this.stateCounter += 1;
    const key = node ?? `state_${this.stateCounter}`;
    const slot = this.store.value(key);
    slot.set(initial);
    slot.markApplied();
    return new State<T>(key, this);
  }

  on(intent: string, handler: (...args: unknown[]) => void): void {
    this.neon.router.on(intent, handler as never);
  }

  async run(): Promise<void> {
    // The Node runtime uses the same blocking serve loop; for now this is a
    // no-op because the NeonApp already serves on its own socket server.
  }

  async publishNow(): Promise<void> {
    try {
      await this.neon.ui.publish();
    } catch {
      // Offline apps / tests may not have a live connection.
    }
  }

  async dispose(): Promise<void> {
    await this.renderer.flush();
    this.neon.stop();
    if (currentApp === this) currentApp = null;
  }
}

// ---------------------------------------------------------------------------
// Module-level helpers.
// ---------------------------------------------------------------------------

export async function start(
  options: Parameters<typeof NeonApp.start>[0] = {},
): Promise<FacadeApp> {
  return FacadeApp.boot(options);
}

export async function mount(source: string): Promise<unknown> {
  return getCurrentApp().mount(source);
}

export function state<T>(initial: T, node?: string): State<T> {
  return getCurrentApp().state(initial, node);
}

export function on(intent: string, handler: (...args: unknown[]) => void): void {
  getCurrentApp().on(intent, handler);
}

// performance.now shim for non-browser runtimes
declare const performance: { now(): number };

File: packages/python-sdk/src/neon3_sdk/facade.py
"""React-hooks-style façade over the Neon3 SDK.

This is the public "easy mode" API. It wraps :class:`~neon3_sdk.app.NeonApp`
and :class:`~neon3_sdk.store.ObservableStore` so a counter demo reads::

    from neon3_sdk.facade import start, mount, state, on, run

    with start(mode="windowed", origin="demo"):
        mount("counter.nui")
        count = state(0)

        @on("btn.increment")
        def _():
            count.value += 1

        run()

The lower-level ``NeonApp`` is still available as ``app.neon`` inside the
context if you need fine-grained control.
"""

from __future__ import annotations

import threading
import time
from pathlib import Path
from typing import Any, Callable, Iterator

from .app import NeonApp
from .constants import AnimationAction
from .errors import CapabilityError
from .store import ObservableStore

# ---------------------------------------------------------------------------
# Global "current app" context (like React's current renderer).
#
# ``with start(...)`` pushes a FacadeApp onto this thread-local stack; the
# module-level ``mount``/``state``/``on`` helpers operate on the top. This
# keeps the example code free of explicit ``app.`` everywhere.
# ---------------------------------------------------------------------------

_thread_local = threading.local()


def _current() -> "FacadeApp":
    try:
        stack: list[FacadeApp] = _thread_local.stack  # type: ignore[attr-defined]
    except AttributeError:
        stack = []
        _thread_local.stack = stack  # type: ignore[attr-defined]
    if not stack:
        raise RuntimeError(
            "No active Neon3 app. Wrap your code in `with start(...) as app:` "
            "before calling mount()/state()/on()."
        )
    return stack[-1]


def _push(app: "FacadeApp") -> None:
    try:
        stack: list[FacadeApp] = _thread_local.stack  # type: ignore[attr-defined]
    except AttributeError:
        stack = []
        _thread_local.stack = stack  # type: ignore[attr-defined]
    stack.append(app)


def _pop() -> None:
    try:
        stack: list[FacadeApp] = _thread_local.stack  # type: ignore[attr-defined]
        stack.pop()
    except (AttributeError, IndexError):
        pass


# ---------------------------------------------------------------------------
# State[T]: a readable/writable scalar that auto-publishes on assignment.
# ---------------------------------------------------------------------------


class State:
    """A readable/writable scalar bound to the current app's store.

    ``s.value`` reads the current value; ``s.value = x`` writes it and
    immediately publishes one ``ui.input.frame`` (throttled to the next
    microtask / RPC batch). This is the ``useState`` analogue.
    """

    __slots__ = ("_key", "_app", "_store")

    def __init__(self, key: str, app: "FacadeApp") -> None:
        self._key = key
        self._app = app
        self._store = app.store

    @property
    def key(self) -> str:
        return self._key

    @property
    def value(self) -> Any:
        raw = self._store.value(self._key).get()
        # Unwrap the wire envelope {"kind": ..., "value": ...} into the raw
        # Python value so users get `42`, not `{"kind": "i32", "value": 42}`.
        if isinstance(raw, dict) and "kind" in raw and "value" in raw:
            return raw["value"]
        return raw

    @value.setter
    def value(self, new_value: Any) -> None:
        self._store.value(self._key).set(new_value)
        self._app._publish_now()

    # Python niceties: let `count += 1` work by reading, modifying, writing.
    def __iadd__(self, other: Any) -> "State":
        self.value = self.value + other  # type: ignore[operator]
        return self

    def __repr__(self) -> str:
        return f"State({self._key!r}={self.value!r})"


# ---------------------------------------------------------------------------
# Sub-namespaces: app.shader.* / app.renderer.* / app.anim.* / app.surface.*
# ---------------------------------------------------------------------------


class _ShaderNamespace:
    """``app.shader.on(event_id)`` — subscribe to WGSL emit_shader_event."""

    def __init__(self, app: "FacadeApp") -> None:
        self._app = app

    def on(self, event_id: int) -> Callable[[Callable[[list[float]], Any]], Callable[[list[float]], Any]]:
        """Decorator: register a handler for a specific shader event_id.

        The handler receives a ``list[float]`` of length 4 (the payload).
        """
        def decorator(handler: Callable[[list[float]], Any]) -> Callable[[list[float]], Any]:
            self._app._shader_handlers[int(event_id)] = handler
            return handler
        return decorator


class _RendererNamespace:
    """``app.renderer.view_extras(...)`` — upload shader uniform data."""

    def __init__(self, app: "FacadeApp") -> None:
        self._app = app
        self._last_flush = 0.0
        self._pending: list[list[float]] | None = None

    def view_extras(self, *rows: list[float]) -> None:
        """Upload 1..10 rows of vec4 to ``view.extras[0..9]``.

        Fewer than 10 rows are padded with ``[0, 0, 0, 0]``. Calls are
        throttled to 60 Hz; multiple calls within one frame batch into a
        single RPC.
        """
        if not 1 <= len(rows) <= 10:
            raise ValueError(f"view_extras expects 1..10 rows, got {len(rows)}")
        for i, row in enumerate(rows):
            if len(row) != 4:
                raise ValueError(f"row {i} must have 4 floats, got {len(row)}")
        self._pending = [list(r) for r in rows]
        now = time.monotonic()
        if now - self._last_flush >= 1.0 / 60.0:
            self.flush()

    def flush(self) -> None:
        """Force-pending view_extras immediately (bypasses the 60 Hz throttle)."""
        if self._pending is None:
            return
        rows = self._pending
        self._pending = None
        self._last_flush = time.monotonic()
        if self._app.neon.render is None:
            return
        self._app.neon.render.set_view_extras(rows)


class _AnimNamespace:
    """``app.anim.pause/resume/seek/cancel`` — renderer-owned timeline control."""

    def __init__(self, app: "FacadeApp") -> None:
        self._app = app

    def _render(self):
        if self._app.neon.render is None:
            raise CapabilityError("wgpu.ui.animation.control.v1")
        return self._app.neon.render

    def pause(self, node_path: str) -> None:
        self._render().animation_pause(node_path)

    def resume(self, node_path: str) -> None:
        self._render().animation_resume(node_path)

    def cancel(self, node_path: str) -> None:
        self._render().animation_cancel(node_path)

    def seek(self, node_path: str, progress: float) -> None:
        self._render().animation_seek(node_path, progress)


# ---------------------------------------------------------------------------
# FacadeApp: the context-manager object returned by start().
# ---------------------------------------------------------------------------


class PNG:
    """A rendered PNG returned by ``app.surface.render(...)``."""

    def __init__(self, path: str) -> None:
        self._path = path

    @property
    def path(self) -> str:
        """The on-disk path the runtime wrote."""
        return self._path

    def save(self, path: str) -> None:
        """Copy the rendered PNG to ``path``."""
        import shutil
        shutil.copyfile(self._path, path)

    def bytes(self) -> bytes:
        """Read the PNG bytes from disk."""
        with open(self._path, "rb") as f:
            return f.read()


class _SurfaceNamespace:
    """``app.surface.render(nui, size=(w, h)) -> PNG`` — offscreen render to file."""

    def __init__(self, app: "FacadeApp") -> None:
        self._app = app

    def render(self, nui: str, *, size: tuple[int, int] = (1280, 720),
               surface_id: str = "facade-offscreen") -> PNG:
        """Render a NUI flow offscreen and return a PNG handle.

        Opens a shared surface, mounts the flow, captures one frame as PNG,
        and returns a :class:`PNG` you can ``.save(path)`` or ``.bytes()``.
        """
        import os
        import uuid
        from .render import ExternalSurface, SurfaceKind, SurfaceOpen, SurfaceSize

        if self._app.neon.render is None:
            raise RuntimeError("render client is not available (offline?)")
        w, h = size
        sid = f"{surface_id}-{uuid.uuid4().hex[:8]}"
        out_path = os.path.join(os.environ.get("TEMP", "."), f"neon3-{uuid.uuid4().hex[:8]}.png")
        surface = self._app.neon.render.open_surface(SurfaceOpen(
            session_id=self._app.origin,
            surface_id=sid,
            kind=SurfaceKind.SCREEN_UI,
            size=SurfaceSize(width=w, height=h),
            buffer_count=1,
        ))
        # Mount flow onto the surface
        self._app.neon.ui.mount_flow(nui)
        surface.save_png(out_path)
        return PNG(out_path)


class FacadeApp:
    """The high-level app object. Created by :func:`start`, used as a context
    manager. Inside ``with start(...) as app:`` the module-level helpers
    (``mount``/``state``/``on``) operate on it.
    """

    def __init__(self, neon: NeonApp) -> None:
        self.neon = neon
        self.store: ObservableStore = neon.store or ObservableStore()
        if neon.store is None:
            neon.store = self.store
        self._state_counter = 0
        self._shader_handlers: dict[int, Callable[[list[float]], Any]] = {}
        self._shader_thread: threading.Thread | None = None
        self._stop_flag = threading.Event()
        # sub-namespaces
        self.shader = _ShaderNamespace(self)
        self.renderer = _RendererNamespace(self)
        self.anim = _AnimNamespace(self)
        self.surface = _SurfaceNamespace(self)

    # --------------------------------------------------------- lifecycle
    def __enter__(self) -> "FacadeApp":
        _push(self)
        self._start_shader_dispatch()
        return self

    def __exit__(self, *_exc: Any) -> None:
        self._stop_flag.set()
        try:
            self.renderer.flush()
        finally:
            try:
                self.neon.stop()
            finally:
                _pop()

    # --------------------------------------------------------- helpers
    def mount(self, source: str | Path) -> Any:
        """Mount a NUI flow. Accepts a file path (``.nui``) or a NUI string."""
        if isinstance(source, (str, Path)) and Path(source).is_file():
            return self.neon.ui.mount_flow_file(source)  # type: ignore[union-attr]
        return self.neon.ui.mount_flow(str(source))  # type: ignore[union-attr]

    def state(self, initial: Any, *, node: str | None = None) -> State:
        """Register a new reactive scalar and return a ``State`` handle."""
        self._state_counter += 1
        key = node or f"state_{self._state_counter}"
        slot = self.store.value(key)
        slot.set(initial)
        slot.mark_applied()
        return State(key, self)

    def on(self, intent: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
        """Decorator: register a semantic-intent handler."""
        return self.neon.intent(intent)

    def run(self, *, block: bool = True) -> Any:
        """Block and serve the domain endpoint (callbacks arrive here)."""
        return self.neon.run(block=block)

    # --------------------------------------------------------- internals
    def _publish_now(self) -> None:
        """Flush pending scalar diffs to the runtime immediately."""
        try:
            self.neon.publish()
        except Exception:
            # Offline apps / tests may not have a live connection; swallow.
            pass

    def _start_shader_dispatch(self) -> None:
        """Background thread: subscribe to shader.event and dispatch by id."""
        if self.neon.events is None:
            return

        def loop() -> None:
            try:
                sub = self.neon.events.subscribe("shader.event")  # type: ignore[union-attr]
            except Exception:
                return
            while not self._stop_flag.is_set():
                try:
                    envelope = sub.recv(timeout=0.2)
                except Exception:
                    continue
                if envelope is None:
                    continue
                payload = envelope.payload
                event_id = int(payload.get("event_id", 0))
                handler = self._shader_handlers.get(event_id)
                if handler is None:
                    continue
                values = [float(v) for v in payload.get("payload", [0.0, 0.0, 0.0, 0.0])]
                try:
                    handler(values)
                except Exception:
                    # A user handler must not kill the dispatch thread.
                    pass

        self._shader_thread = threading.Thread(target=loop, daemon=True)
        self._shader_thread.start()


# ---------------------------------------------------------------------------
# Module-level helpers (operate on the current app).
# ---------------------------------------------------------------------------


def start(
    *,
    mode: str = "windowed",
    origin: str = "neon3-facade",
    **kwargs: Any,
) -> FacadeApp:
    """Boot (or attach to) the Neon3 runtime and return a context manager.

    Use as ``with start(...) as app: ...``. The app is automatically stopped
    when the block exits.
    """
    neon = NeonApp.start(mode=mode, origin=origin, **kwargs)
    return FacadeApp(neon)


def mount(source: str | Path) -> Any:
    """Mount a NUI flow on the current app."""
    return _current().mount(source)


def state(initial: Any, *, node: str | None = None) -> State:
    """Create a reactive scalar bound to the current app's store."""
    return _current().state(initial, node=node)


def on(intent: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
    """Register a semantic-intent handler on the current app."""
    return _current().on(intent)


def run(*, block: bool = True) -> Any:
    """Block and serve (only needed after the ``with`` block if you want to
    keep the process alive; inside the block this is optional)."""
    return _current().run(block=block)
